numpyro/util.py

```diff
@@ -309,3 +311,263 @@ def soft_vmap(fn, xs, batch_ndims=1, chunk_size=None):
     map_ndims = int(num_chunks > 1) + int(chunk_size > 1)
     ys = tree_map(lambda y: jnp.reshape(y, (-1,) + jnp.shape(y)[map_ndims:])[:batch_size], ys)
     return tree_map(lambda y: jnp.reshape(y, batch_shape + jnp.shape(y)[1:]), ys)
+
+
+def get_model_relations(model, model_args=None, model_kwargs=None, num_tries=10):
+    """
+    Infer relations of RVs and plates from given model and optionally data.
+    See issue #949 on pyro-ppl/numpyro for more details.
+
+    :param int num_tries: times to trace model to detect discrete -> continuous dependency.
+    """
+    # TODO: put import in more sensible location
+    from numpyro import handlers
+    import numpyro.distributions as dist
```

Comment:
@kpj Per discussion with other devs, it is better to move the implementation to a separate file `numpyro/render.py` or `numpyro/contrib/render.py`. This will resolve the circular import issue and preserve the independence of `numpyro.util`.

Reply:
Either option is fine by me. I assume the general consensus is to prefer the second one (moving everything to `numpyro.contrib.render` and keeping the `graphviz` dependency in `extras_require.dev`)?

Reply:
You are right!